File: SexyBiscuit.Engine/Core/Bounds.cs
using Microsoft.Xna.Framework;

namespace SexyBiscuit.Engine.Core;

/// <summary>
/// An axis-aligned bounding box stored as centre plus half-extents.
/// Used for frustum culling, broad-phase queries and LOD sizing.
/// </summary>
public struct Bounds : IEquatable<Bounds>
{
    /// <summary>Centre of the box in world (or local) space.</summary>
    public Vector3 Center;

    /// <summary>Half the size of the box along each axis. Always non-negative.</summary>
    public Vector3 Extents;

    /// <summary>Full size of the box along each axis.</summary>
    public Vector3 Size
    {
        readonly get => Extents * 2f;
        set => Extents = new Vector3(MathF.Abs(value.X), MathF.Abs(value.Y), MathF.Abs(value.Z)) * 0.5f;
    }

    /// <summary>Corner of the box with the smallest coordinate on every axis.</summary>
    public readonly Vector3 Min => Center - Extents;

    /// <summary>Corner of the box with the largest coordinate on every axis.</summary>
    public readonly Vector3 Max => Center + Extents;

    /// <summary>Radius of the sphere that fully contains this box.</summary>
    public readonly float BoundingRadius => Extents.Length();

    public Bounds(Vector3 center, Vector3 size)
    {
        Center  = center;
        Extents = new Vector3(MathF.Abs(size.X), MathF.Abs(size.Y), MathF.Abs(size.Z)) * 0.5f;
    }

    /// <summary>A degenerate box at the origin. Grow it with <see cref="Encapsulate(Vector3)"/>.</summary>
    public static Bounds Zero => new(Vector3.Zero, Vector3.Zero);

    /// <summary>Builds the tightest box containing every point in <paramref name="points"/>.</summary>
    public static Bounds FromPoints(IEnumerable<Vector3> points)
    {
        var min = new Vector3(float.MaxValue);
        var max = new Vector3(float.MinValue);
        bool any = false;

        foreach (var p in points)
        {
            any = true;
            min = Vector3.Min(min, p);
            max = Vector3.Max(max, p);
        }

        if (!any) return Zero;
        return FromMinMax(min, max);
    }

    /// <summary>Builds a box from opposite corners.</summary>
    public static Bounds FromMinMax(Vector3 min, Vector3 max)
        => new() { Center = (min + max) * 0.5f, Extents = (max - min) * 0.5f };

    /// <summary>Grows the box so that it contains <paramref name="point"/>.</summary>
    public void Encapsulate(Vector3 point)
        => this = FromMinMax(Vector3.Min(Min, point), Vector3.Max(Max, point));

    /// <summary>Grows the box so that it contains <paramref name="other"/>.</summary>
    public void Encapsulate(Bounds other)
        => this = FromMinMax(Vector3.Min(Min, other.Min), Vector3.Max(Max, other.Max));

    /// <summary>Expands the box by <paramref name="amount"/> on every side.</summary>
    public void Expand(float amount) => Extents += new Vector3(amount);

    public readonly bool Contains(Vector3 p)
    {
        var min = Min; var max = Max;
        return p.X >= min.X && p.X <= max.X
            && p.Y >= min.Y && p.Y <= max.Y
            && p.Z >= min.Z && p.Z <= max.Z;
    }

    public readonly bool Intersects(Bounds other)
    {
        var aMin = Min; var aMax = Max;
        var bMin = other.Min; var bMax = other.Max;
        return aMin.X <= bMax.X && aMax.X >= bMin.X
            && aMin.Y <= bMax.Y && aMax.Y >= bMin.Y
            && aMin.Z <= bMax.Z && aMax.Z >= bMin.Z;
    }

    /// <summary>Converts to a MonoGame <see cref="BoundingBox"/> for use with its intersection helpers.</summary>
    public readonly BoundingBox ToBoundingBox() => new(Min, Max);

    /// <summary>The smallest sphere fully containing this box.</summary>
    public readonly BoundingSphere ToBoundingSphere() => new(Center, BoundingRadius);

    /// <summary>
    /// Transforms the box by <paramref name="matrix"/> and returns the axis-aligned box
    /// containing the result. The returned box is conservative — it may be larger than
    /// the tightest fit when the matrix contains rotation.
    /// </summary>
    public readonly Bounds Transform(Matrix matrix)
    {
        // Transform the centre, then accumulate the extents against the absolute
        // value of the rotation/scale basis — the standard conservative AABB transform.
        var center = Vector3.Transform(Center, matrix);

        var extents = new Vector3(
            MathF.Abs(matrix.M11) * Extents.X + MathF.Abs(matrix.M21) * Extents.Y + MathF.Abs(matrix.M31) * Extents.Z,
            MathF.Abs(matrix.M12) * Extents.X + MathF.Abs(matrix.M22) * Extents.Y + MathF.Abs(matrix.M32) * Extents.Z,
            MathF.Abs(matrix.M13) * Extents.X + MathF.Abs(matrix.M23) * Extents.Y + MathF.Abs(matrix.M33) * Extents.Z);

        return new Bounds { Center = center, Extents = extents };
    }

    /// <summary>Closest point on or inside the box to <paramref name="point"/>.</summary>
    public readonly Vector3 ClosestPoint(Vector3 point) => Vector3.Clamp(point, Min, Max);

    /// <summary>Shortest distance from <paramref name="point"/> to the box surface; 0 when inside.</summary>
    public readonly float DistanceTo(Vector3 point) => Vector3.Distance(ClosestPoint(point), point);

    public readonly bool Equals(Bounds other) => Center == other.Center && Extents == other.Extents;
    public readonly override bool Equals(object? obj) => obj is Bounds b && Equals(b);
    public readonly override int GetHashCode() => HashCode.Combine(Center, Extents);
    public static bool operator ==(Bounds a, Bounds b) => a.Equals(b);
    public static bool operator !=(Bounds a, Bounds b) => !a.Equals(b);

    public readonly override string ToString() => $"Bounds(centre {Center}, size {Size})";
}

File: html5/src/math/Bounds.js
// -----------------------------------------------------------------------------
// Rect and Bounds — the 2D and 3D axis-aligned volumes the engine tests against.
// -----------------------------------------------------------------------------

import { Vector2 } from './Vector2.js';
import { Vector3 } from './Vector3.js';

/** An axis-aligned 2D rectangle, positioned by its top-left corner. */
export class Rect {
    constructor(x = 0, y = 0, width = 0, height = 0) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    get left()   { return this.x; }
    get right()  { return this.x + this.width; }
    get top()    { return this.y; }
    get bottom() { return this.y + this.height; }

    get center() { return new Vector2(this.x + this.width / 2, this.y + this.height / 2); }
    get size()   { return new Vector2(this.width, this.height); }

    clone() { return new Rect(this.x, this.y, this.width, this.height); }

    contains(point) {
        return point.x >= this.x && point.x <= this.right
            && point.y >= this.y && point.y <= this.bottom;
    }

    /** True when the two rectangles overlap. Touching edges count as overlapping. */
    intersects(other) {
        return this.x <= other.right && this.right >= other.x
            && this.y <= other.bottom && this.bottom >= other.y;
    }

    /** The overlapping region, or null when they do not overlap. */
    intersection(other) {
        const x = Math.max(this.x, other.x);
        const y = Math.max(this.y, other.y);
        const r = Math.min(this.right, other.right);
        const b = Math.min(this.bottom, other.bottom);
        if (r < x || b < y) return null;
        return new Rect(x, y, r - x, b - y);
    }

    /** The smallest rectangle containing both. */
    union(other) {
        const x = Math.min(this.x, other.x);
        const y = Math.min(this.y, other.y);
        return new Rect(x, y,
            Math.max(this.right, other.right) - x,
            Math.max(this.bottom, other.bottom) - y);
    }

    /** Grows the rectangle by `amount` on every side. */
    inflate(amount) {
        return new Rect(
            this.x - amount, this.y - amount,
            this.width + amount * 2, this.height + amount * 2);
    }

    /** A rectangle from its centre and full size, the form colliders use. */
    static fromCenter(center, size) {
        return new Rect(center.x - size.x / 2, center.y - size.y / 2, size.x, size.y);
    }

    toJSON() { return [this.x, this.y, this.width, this.height]; }
    toString() { return `Rect(${this.x}, ${this.y}, ${this.width}x${this.height})`; }
}

/** An axis-aligned 3D box, stored as centre and half-extents. */
export class Bounds {
    /**
     * @param {Vector3} [center] Centre of the box.
     * @param {Vector3} [size]   Full size on each axis, not half-extents.
     */
    constructor(center = new Vector3(), size = new Vector3()) {
        this.center = Vector3.from(center);
        this.size = Vector3.from(size);
    }

    get extents() { return Vector3.scale(this.size, 0.5); }
    get min() { return Vector3.subtract(this.center, this.extents); }
    get max() { return Vector3.add(this.center, this.extents); }

    clone() { return new Bounds(this.center.clone(), this.size.clone()); }

    contains(point) {
        const min = this.min, max = this.max;
        return point.x >= min.x && point.x <= max.x
            && point.y >= min.y && point.y <= max.y
            && point.z >= min.z && point.z <= max.z;
    }

    intersects(other) {
        const aMin = this.min, aMax = this.max;
        const bMin = other.min, bMax = other.max;
        return aMin.x <= bMax.x && aMax.x >= bMin.x
            && aMin.y <= bMax.y && aMax.y >= bMin.y
            && aMin.z <= bMax.z && aMax.z >= bMin.z;
    }

    /** Grows the box just enough to contain `point`. */
    encapsulate(point) {
        const min = this.min, max = this.max;
        const newMin = new Vector3(
            Math.min(min.x, point.x), Math.min(min.y, point.y), Math.min(min.z, point.z));
        const newMax = new Vector3(
            Math.max(max.x, point.x), Math.max(max.y, point.y), Math.max(max.z, point.z));
        return Bounds.fromMinMax(newMin, newMax);
    }

    /** The point inside the box nearest to `point`. */
    closestPoint(point) {
        const min = this.min, max = this.max;
        return new Vector3(
            Math.max(min.x, Math.min(point.x, max.x)),
            Math.max(min.y, Math.min(point.y, max.y)),
            Math.max(min.z, Math.min(point.z, max.z)));
    }

    /** The radius of the sphere that fully contains this box. */
    get boundingRadius() { return this.extents.length; }

    static fromMinMax(min, max) {
        const size = Vector3.subtract(max, min);
        return new Bounds(Vector3.add(min, Vector3.scale(size, 0.5)), size);
    }

    toString() { return `Bounds(center=${this.center}, size=${this.size})`; }
}
